types/prefs/prefs.go

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package prefs contains types and functions to work with arbitrary
+// preference hierarchies.
+//
+// Specifically, the package provides [Item], [List], [Map], [StructList] and [StructMap]
+// types which represent individual preferences in a user-defined prefs struct.
+// A valid prefs struct must contain one or more exported fields of the preference types,
+// either directly or within nested structs, but not pointers to these types.
+// Additionally to preferences, a prefs struct may contain any number of
+// non-preference fields that will be marshalled and unmarshalled but are
+// otherwise ignored by the prefs package.
+//
+// The preference types are compatible with the [tailscale.com/cmd/viewer] and
+// [tailscale.com/cmd/cloner] utilities. It is recommended to generate a read-only view
+// of the user-defined prefs structure and use it in place of prefs whenever the prefs
+// should not be modified.
+package prefs
+
+import (
+	"errors"
+
+	jsonv2 "github.com/go-json-experiment/json"
+	"github.com/go-json-experiment/json/jsontext"
+	"tailscale.com/types/opt"
+)
+
+var (
+	// ErrManaged is the error returned when attempting to modify a managed preference.
+	ErrManaged = errors.New("cannot modify a managed preference")
+	// ErrReadOnly is the error returned when attempting to modify a readonly preference.
+	ErrReadOnly = errors.New("cannot modify a readonly preference")
+)
+
+// metadata holds type-agnostic preference metadata.
+type metadata struct {
+	// Managed indicates whether the preference is managed via MDM, Group Policy, or other means.
+	Managed bool `json:",omitzero"`
+
+	// ReadOnly indicates whether the preference is read-only due to any other reasons,
+	// such as user's access rights.
+	ReadOnly bool `json:",omitzero"`
+}
+
+// serializable is a JSON-serializable preference data.
+type serializable[T any] struct {
+	// Value is an optional preference value that is set when the preference is
+	// configured by the user or managed by an admin.
+	Value opt.Value[T] `json:",omitzero"`
+	// Default is the default preference value to be used
+	// when the preference has not been configured.
+	Default T `json:",omitzero"`
```

Comment:
should Default be in the serialized version, or perhaps in `preference`?

Reply:
It's in the serialized version, so that we can pass it from the backend to the client (GUI, CLI or web).
I believe the defaults should be maintained on the `tailscaled` side, so that the clients can use the fetched preferences as-is, without having to maintain their own defaults throughout the codebase or modify the preferences received from the backend before utilizing them.